src/structures/AuditLog.ts

```diff
@@ -0,0 +1,347 @@
+  pruneDeleteDays: number;
+  widgetEnabled: boolean;
+  widgetChannelId: Snowflake;
+  systemChannelId: Snowflake;
+  position: number;
+  topic: string;
+  bitrate: number;
+  permissionOverwrites: PermissionOverwrite[];
+  nsfw: boolean;
+  application_id: Snowflake;
+  rateLimitPerUser: number;
+  permissions: number;
+  permissionsNew: string;
+  color: number;
+  hoist: boolean;
+  mentionable: boolean;
+  allow: number;
+  allowNew: string;
+  deny: number;
+  denyNew: string;
+  code: string;
+  channelId: Snowflake;
+  inviterId: Snowflake;
+  maxUses: number;
+  uses: number;
+  maxAge: number;
+  temporary: boolean;
+  deaf: boolean;
+  mute: boolean;
+  nick: string;
+  avatarHash: string;
+  id: Snowflake;
+  type: number | string;
+  enableEmoticons: boolean;
+  expireBehavior: number;
+  expireGracePeriod: number;
+}
+
+export interface UnspecificChange<T extends keyof ChangeKey> {
+  newValue?: ChangeKey[T];
+  oldValue?: ChangeKey[T];
+  key: T;
+}
+
+type SpecificChange<T extends keyof ChangeKey> = T extends keyof ChangeKey
+  ? UnspecificChange<T>
+  : never;
+
+export type Change = SpecificChange<keyof ChangeKey>;
+
+export function parseAuditLog(
+  client: Client,
+  { audit_log_entries, integrations, users, webhooks }: auditLog.AuditLog,
+): AuditLog {
+  return {
+    users: users.map((user) => new User(client, user)),
+    entries: audit_log_entries.map((entry) => parseEntry(entry)),
+    integrations,
+    webhooks: webhooks.map((webhook) => parseWebhook(client, webhook)),
+  };
+}
+
+function parseEntry(entry: auditLog.Entry): Entry {
+  let extra;
+
+  if (entry.options) {
+    switch (entry.action_type) {
+      case 13:
+      case 14:
+      case 15:
+        extra = {
+          id: entry.options.id,
+          type: entry.options.type,
+          roleName: entry.options.role_name,
+        } as ChannelOverwriteMember | ChannelOverwriteRole;
+        break;
+      case 21:
+        extra = {
+          deleteMemberDays: entry.options.delete_member_days,
+          membersRemoved: entry.options.members_removed,
+        };
+        break;
+      case 26:
+      case 72:
+        extra = {
+          channelId: entry.options.channel_id,
+          count: entry.options.count,
+        };
+        break;
+      case 27:
+      case 73:
+        extra = {
+          count: entry.options.count,
+        };
+        break;
+      case 74:
+      case 75:
+        extra = {
+          channelId: entry.options.channel_id,
+          messageId: entry.options.message_id,
+        };
+        break;
+    }
+  }
+
+  return {
+    targetId: entry.target_id,
+    changes: entry.changes?.map((change) => {
+      return {
+        key: changeKeyMap[change.key],
+        newValue: change.new_value,
+        oldValue: change.old_value,
+      };
+    }),
+    userId: entry.user_id,
+    id: entry.id,
+    actionType: actionTypeMap[entry.action_type],
+    extra: extra,
+    reason: entry.reason,
+  } as Entry;
```

Comment:
Consider setting explicit return type to `Entry` to avoid mistakes that type casting can introduce.